slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -92,6 +92,52 @@ ::grpc::Status SLAMServiceImpl::GetPositionNew(
     return grpc::Status::OK;
 }
 
+::grpc::Status SLAMServiceImpl::GetPointCloudMap(
+    ServerContext *context, const GetPointCloudMapRequest *request,
+    GetPointCloudMapResponse *response) {
+    std::string pointcloud_map;
+    // Write or grab the latest pointcloud map in form of a string
+    try {
+        std::shared_lock optimization_lock{optimization_shared_mutex,
+                                           std::defer_lock};
+        if (optimization_lock.try_lock()) {
+            // We are able to lock the optimization_shared_mutex, which means
+            // that the optimization is not ongoing and we can grab the newest
+            // map
+            GetLatestSampledPointCloudMapString(pointcloud_map);
+        } else {
+            // We couldn't lock the mutex which means the optimization process
+            // locked it and we need to use the backed up latest map
+            std::lock_guard<std::mutex> lk(viam_response_mutex);
+            pointcloud_map = latest_pointcloud_map;
+        }
+    } catch (std::exception &e) {
+        LOG(ERROR) << "error creating pcd map: " << e.what();
+        std::ostringstream oss;
+        viam::b_continue_session = false;
```

Comment:
Setting `b_continue_session` here. Currently, this is also used in [main.cc](https://github.com/viamrobotics/slam/blob/main/slam-libraries/viam-cartographer/src/main.cc#L15). I think atomic bool is supposed to be safe but I am going from memory

Reply:
Yes, it's safe to concurrently access an atomic bool. However, I don't think we should the normal shutdown procedure here, since the state is corrupted. I think you should just re-throw the exception and let that terminate the process, so we won't do any additional work. Alternatively, if you're worried about callers catching the exception, you can use std::terminate.

Reply:
`throw` was not working within the grpc functions for some reason, but std::terminate seems to work! Swapping to that